### PR TITLE
Adds turnstiles

### PR DESCRIPTION
**Pull request by AffectedArc07**
**What does this PR do:**
This PR adds turnstiles. These are similar to airlocks but can only be walked through one way. Port of old Oracle code. *F*. I have added them to the HoP line and Brig exit. These can have a required access set to them, and will deny if that is not matched. A person with access can drag a person without access through a turnstile (Example: An engineer could drag a scientist through an engineering locked turnstile, but the scientist could not go through on his own)

Walking through a turnstile with access:
d3p0dw7klp

Turnstile with no access:
z36frqh91s

Walking through a turnstile in any direction other than the mapped one will notify the user in chat, and refuse to let them pass
image

It also doesnt insta-clog the chat
q31s0vrkrx

**Images of sprite/map changes:**
Brig: Turnstiles have been added so prisoners can leave without bugging security or the AI. They cannot leave mid sentence due to the cell doors being closed. A second turnstile is there to allow exit from the "poverty cell".
image
image

A single turnstile has been added at the hop line to stop annoying ass tailgaters from cramming their way into the front of the line with people there. One has not been added at the bottom so that people can still leave without having to barge out the front.
image

**Changelog:**
:cl: AffectedArc07
add: Turnstiles are now in the game! For those who do not know, they are like airlocks but only work 1 way
add: Turnstiles have been added at the Brig and HoP line
imagedel: Turnstile sprites
/:cl:

### CODE REVIEW

**Comment by Ralta** [comment]
Does this also include the top set of brig doors? Or just the exit by the cells?

**Comment by Superhats** [comment]
How do you build a turnstile? Can you build a turnstile?

**Comment by AffectedArc07** [comment]
@Ralta Only by the cells, I didnt feel like they were needed at the top sets

@Superhats at the moment, you cannot

Also mapmerge seems to be broken. I did mapmerge this but apparently that didnt work

**Comment by datlo** [comment]
The turnstiles you've mapped in don't have access right? Otherwise they're not much use since they still block prisonners from leaving.

**Comment by AffectedArc07** [comment]
@datlo No they do not

**Comment by taukausanake** [comment]
I like this, but we need to be able to construct and deconstruct these before they are mapped in I feel. Also kind of weird that something that is traditionally mechanical (read: only moving parts) is made with an electronic function. I'm not against the ID blocking but I'd like a variant that is not that and don't just say "make one that doesn't have ID restrictions". I'd rather have the bare bone structure and then add the restrictions to it. Maybe an airlock board or something can be attached. That's just me though

File: paradise.dme [hunk 1]
```diff
@@ -1,2383 +1,2384 @@
-// DM Environment file for paradise.dme.
```
Comment (AffectedArc07) [inline paradise.dme]: WHAT THE FUCK IS WRONG WITH THIS WHY WONT MY MERGE WORK

**Comment by AffectedArc07** [comment]
I mean @taukausanake every turnstile I have seen IRL has some sort of electronic attachment to it. At london they have keycards to get into the stations, at stadiums they have an operator to let you in or out. Every instance I have seen these electronic.

**Comment by taukausanake** [comment]
Not the ones I've seen. Maybe an electronic lock but not an ID thing. The city zoo has them for exits but you just walk out them

**Comment by Spartan6** [comment]
@AffectedArc07 Couple questions.

1. Are they hackable?
2. Are they hooked into the object damage system like normal airlocks?

**Comment by AffectedArc07** [comment]
@Spartan6 
1: No
2: No

**Comment by Spartan6** [comment]
Well the second and really the first should be a thing.

**Comment by AffectedArc07** [comment]
@Spartan6 If we go of what taukau said, a mechanical thing shouldnt be able to be backed. In terms of object damage, I am kinda on the line, as its gonna be hard to destroy something made of steel
image
Like I dont see you caving this in with a toolbox

**Comment by Spartan6** [comment]
@AffectedArc07 I was meaning the 21 damage cap, the same as airlocks.

In regards to the hacking, you have electronic ones. Those should be able to be hacked. You could just give them mostly the same wires as the airlocks. Security ratings too can be added from airlocks.

**Comment by Citinited** [comment]
Is the turnstile at the HoP's line really necessary? I feel like this is easily fixable by everyone ganging up on the person butting in rather than needing anything mapped in to fix it. Plus, the HoP gets a wall-mounted flash for a reason; and this will help prevent opportunistic shenanigans that otherwise wouldn't take place (instead of being able to get to the open windoor in a second, you now have to spend five getting to that point)
.
Also it makes the line feel like a prison. But, that may be the intent.

**Comment by Fox-McCloud** [comment]
Those sprites are hideous.

File: code/game/machinery/turnstile.dm [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	desc = "A mechanical door that permits one-way access and prevents tailgating."
+	icon = 'icons/obj/turnstile.dmi'
+	icon_state = "turnstile_map"
+	power_channel = ENVIRON
+	density = FALSE
+	armor = list(melee = 50, bullet = 50, laser = 50, energy = 50, bomb = 10, bio = 100, rad = 100, fire = 90, acid = 70)
+	anchored = TRUE
+	use_power = FALSE
+	idle_power_usage = 2
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: `INDESTRUCTIBLE`

This is absolutely something that shouldn't be indestructible.

Not even sure this flag does anything on this codebase.

File: code/game/machinery/turnstile.dm [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+/obj/machinery/turnstile/CanAtmosPass(turf/T)
+	return TRUE // Let the air through
+
+/obj/machinery/turnstile/bullet_act(obj/item/projectile/P, def_zone)
+	return -1 // Let those bullets trough, dont ask why this is -1
+
+// Allow access for people being pulled by someone with access
+/obj/machinery/turnstile/proc/allowed_access(var/mob/B)
+	if(B.pulledby && ismob(B.pulledby))
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: pulling people is ok, but grabbing them is not?

File: code/game/machinery/turnstile.dm [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+			if(allowed_access)
+				flick("operate", src)
+				playsound(src,'sound/items/ratchet.ogg',50,0,3)
+				return TRUE
+			else
+				flick("deny", src)
+				playsound(src,'sound/machines/deniedbeep.ogg',50,0,3)
+				return FALSE
+	if(ispath(AM, /obj/item/))
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: `/obj/item/`

Stop this now `/`.

File: code/game/machinery/turnstile.dm [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+				to_chat(usr, "<span class='notice'>\the [src] resists your efforts.</span>")
+				return FALSE
+
+			if(allowed_access)
+				flick("operate", src)
+				playsound(src,'sound/items/ratchet.ogg',50,0,3)
+				return TRUE
+			else
+				flick("deny", src)
+				playsound(src,'sound/machines/deniedbeep.ogg',50,0,3)
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: Spacing.

File: code/game/machinery/turnstile.dm [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+				return FALSE
+			M.last_bumped = world.time
+
+			var/allowed_access = FALSE
+			var/turf/behind = get_step(src, dir)
+
+			if(B in behind.contents)
+				allowed_access = allowed_access(B)
+			else
+				to_chat(usr, "<span class='notice'>\the [src] resists your efforts.</span>")
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: No `\the`. You should know this by now.

File: code/game/machinery/turnstile.dm [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+			else
+				flick("deny", src)
+				playsound(src,'sound/machines/deniedbeep.ogg',50,0,3)
+				return FALSE
+	if(ispath(AM, /obj/item/))
+		return TRUE
+	else
+		return FALSE
+
+/obj/machinery/turnstile/CheckExit(atom/movable/AM as mob|obj, target)
```
Comment (Fox-McCloud) [inline code/game/machinery/turnstile.dm]: No `as X`.

**Comment by Ralta** [comment]
I really do like turnstiles, but I they should probably have a means of construction/deconstruction, as well as hacking. At the very least, they should be destructible.

In any case, it would be lovely for these to get added; they make the Brig much easier to walk in and out of. I would prefer for there to be proper in/out turnstiles at the upper brig entrance though, as it prevents tailgating.

File: code/game/machinery/turnstile.dm (hunk 4 above)
Comment (Citinited) [reply code/game/machinery/turnstile.dm]: also `isobj(AM)`

File: code/game/machinery/turnstile.dm [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	icon_state = "turnstile"
+
+/obj/machinery/turnstile/CanAtmosPass(turf/T)
+	return TRUE // Let the air through
+
+/obj/machinery/turnstile/bullet_act(obj/item/projectile/P, def_zone)
+	return -1 // Let those bullets trough, dont ask why this is -1
+
+// Allow access for people being pulled by someone with access
+/obj/machinery/turnstile/proc/allowed_access(var/mob/B)
```
Comment (Citinited) [inline code/game/machinery/turnstile.dm]: `var/` is not required here

File: code/game/machinery/turnstile.dm (hunk 2 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: Probably does otherwise it would have errored on compile, but will remove

File: code/game/machinery/turnstile.dm (hunk 3 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: Is there a way to check for grabs easily without even more snowflake code

File: code/game/machinery/turnstile.dm (hunk 5 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: What about it?

File: code/game/machinery/turnstile.dm (hunk 6 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: What else do I use, otherwise it just says `turnstile resists your efforts`. Do I use `\improper` in the item name?

File: code/game/machinery/turnstile.dm (hunk 7 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: Is this telling me to use `as X`? Because I dont see an instance of `as X` here

**Comment by AffectedArc07** [comment]
@Fox-McCloud If you can make better sprites I am all for using those instead of the current ones

**Comment by davidchan13** [comment]
While I like the idea, I'm skeptical if these are appropriate in their current state.

Are these solid in regards to projectiles? (Can a laser pass through? A bullet? A thrown grenade?)

Is there anyway to defeat one? I feel the more traditional subway style would be better subway1
Allowing for people to jump a turnstile from the wrong direction as they would climb over a table. This would also allow for code to treat them like tables so things could be thrown or shot over them, as I suspect the current implementation of these puts a bullet proof window effectively in front of the HoP's desk.

TL;DR
:+1: Somewhat sensible 1 way traffic item.
:-1: Not Constructible/Deconstructible.
:-1: Indestructible.
:-1: Not Hackable.

File: code/game/machinery/turnstile.dm (hunk 5 above)
Comment (FalseIncarnate) [reply code/game/machinery/turnstile.dm]: You lack it, so add some spaces after the commas.

File: code/game/machinery/turnstile.dm (hunk 7 above)
Comment (FalseIncarnate) [reply code/game/machinery/turnstile.dm]: `as mob|obj` is pretty clearly there. Fox is saying he doesn't want it there. Remove that bit.

**Comment by FalseIncarnate** [comment]
@davidchan13 it appears that they should allow bullets to pass based on the code, not sure about grenades/thrown objects since those are handled differently. Leaning towards a no on that part.

**Comment by Ty-Omaha** [comment]
Maybe @Ionward could sprite some?

File: code/game/machinery/turnstile.dm (hunk 3 above)
Comment (Fox-McCloud) [reply code/game/machinery/turnstile.dm]: nope.

**Comment by Citinited** [comment]
It should at the very least be affected by explosions. Imagine if someone blew up the brig and all that was left was the turnstiles 😄

File: code/game/machinery/turnstile.dm (hunk 3 above)
Comment (AffectedArc07) [reply code/game/machinery/turnstile.dm]: WElp

**Comment by Ozewse** [comment]
I like this.
I think it's a great idea, given that the execution includes decon/construction and firing through them- although a re-sprite could make them appear more defensive. Like a revolving door instead of a turnstile. 
Also ... could you imagine hacking them causing them to spin super fast and do a fuckload of brute damage to anyone going through?

**Comment by Fox-McCloud** [comment]
These should be buidable, hackable, and deconstructable. They also shouldn't be immune to damage.

Admittedly, I do have some concerns that this is going to encourage officers to throw people in the brig with even less care than before, since it becomes "just leave and forget about it, since they can get out on their own now" (one of the reasons I never implemented Goon's brig ejector). 

These probably have some use...though I really don't think the HoP line is one of them.

**Comment by Landerlow** [comment]
I really think this would be a very nice addition. Especially for the brig area. Not for the HOP as security is sometimes forced to take people away/enter the line from the other side, for example.

**Comment by AffectedArc07** [comment]
When mapmerge2 happens, this will be continued further, as for now this is gonna be on a hiatus, but not abandoned like some other PRs

**Comment by Allfd** [comment]
Not keen on the sprites, or the indestructible nature.  Or that they can't be hacked like a regular airlock.

**Comment by AffectedArc07** [comment]
Oh yeah I have no idea how construction works so someone is gonna need to help me. 

*please help me*

**Comment by davidchan13** [comment]
Not really a coder but dumb idea: Make turnstiles a subtype or child of regular air locks so that they would inherit any future changes to doors? I assume that would bring constructability, hacking and destruction, allow use of air lock electronics to build them (as airlocks and windoors use) and would just require the single direction passing to be added to the subtype? I assume for construction you would either be making an airlock frame and adding rods to it instead of glass to convert it to a turnstile, otherwise you'd have to come up with another machine frame.

The main issue I forsee with constructable turnstiles is that if a turnstile is built in an open corridor, and say it's set up for north to south travel, blocking travel from the south, what happens if someone approaches from the east or west? What happens if someone enters from the north, but doesn't go south but rather goes east or west? Does it block these movements or is it only going to block movements from the south?

Part of me is starting to wonder if it would just be better codewise and gameplay wise to make these an alternate mode of holobarriers or create a holo-turnstile projector that can place these down.

**Comment by AffectedArc07** [comment]
@davidchan13 I thought about that but making a subtype but that would be a lot more effort than it is worth to interrupt mid construction.

In terms of the walking through, lets picture a turnstile that enters north and exits south. You cannot enter east, west or south. If you are on the same tile, you can only exit south, you cannot walk back out the entrance, or out the side

**Comment by AffectedArc07** [comment]
Closed, but anyone else is welcome to take this up.
I personally dont want to touch construction code, feel free to pull my branch if you wish
